### PR TITLE
8252204: AArch64: Implement SHA3 accelerator/intrinsic

### PR DESCRIPTION
**Pull request by RealFYang**
Contributed-by: ard.biesheuvel@linaro.org, dongbo4@huawei.com

This added an intrinsic for SHA3 using aarch64 v8.2 SHA3 Crypto Extensions.
Reference implementation for core SHA-3 transform using ARMv8.2 Crypto Extensions:
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm64/crypto/sha3-ce-core.S?h=v5.4.52

Trivial adaptation in SHA3. implCompress is needed for the purpose of adding the intrinsic.
For SHA3, we need to pass one extra parameter "digestLength" to the stub for the calculation of block size.
"digestLength" is also used in for the EOR loop before keccak to differentiate different SHA3 variants.

We added jtreg tests for SHA3 and used QEMU system emulator which supports SHA3 instructions to test the functionality.
Patch passed jtreg tier1-3 tests with QEMU system emulator.
Also verified with jtreg tier1-3 tests without SHA3 instructions on aarch64-linux-gnu and x86_64-linux-gnu, to make sure that there's no regression.

We used one existing JMH test for performance test: test/micro/org/openjdk/bench/java/security/MessageDigests.java
We measured the performance benefit with an aarch64 cycle-accurate simulator.
Patch delivers 20% - 40% performance improvement depending on specific SHA3 digest length and size of the message.

For now, this feature will not be enabled automatically for aarch64. We can auto-enable this when it is fully tested on real hardware.  But for the above testing purposes, this is auto-enabled when the corresponding hardware feature is detected.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x32 | Linux x64 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |  ✔️ (9/9 passed) | ❌ (1/9 failed) | ✔️ (9/9 passed) |

**Failed test task**
- [Windows x64 (hs/tier1 compiler)](https://github.com/RealFYang/jdk/runs/1289917904)

### Issue
 * [JDK-8252204](https://bugs.openjdk.java.net/browse/JDK-8252204): AArch64: Implement SHA3 accelerator/intrinsic

### Reviewers
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to a5810df62495f3c60749a25aa601daa2fd739857
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Contributors
 * Ard Biesheuvel `<ard.biesheuvel@linaro.org>`
 * Dong Bo `<dongbo4@huawei.com>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/207/head:pull/207`
`$ git checkout pull/207`

### CODE REVIEW

**Comment by RealFYang** [comment]
Bug: https://bugs.openjdk.java.net/browse/JDK-8252204
This added an intrinsic for SHA3 using aarch64 v8.2 SHA3 Crypto Extensions.
Reference implementation for core SHA-3 transform using ARMv8.2 Crypto Extensions:
        https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm64/crypto/sha3-ce-core.S?h=v5.4.52

Trivial adaptation in SHA3. implCompress is needed for the purpose of adding the intrinsic.
For SHA3, we need to pass one extra parameter "digestLength" to the stub for the calculation of block size.
"digestLength" is also used in for the EOR loop before keccak to differentiate different SHA3 variants.

We added jtreg tests for SHA3 and used QEMU system emulator which supports SHA3 instructions to test the functionality. 
Patch passed jtreg tier1-3 tests with QEMU system emulator. 
Also verified with jtreg tier1-3 tests without SHA3 instructions on aarch64-linux-gnu and x86_64-linux-gnu, to make sure that there's no regression. 

We used one existing JMH test for performance test: test/micro/org/openjdk/bench/java/security/MessageDigests.java 
We measured the performance benefit with an aarch64 cycle-accurate simulator.  
Patch delivers 20% - 40% performance improvement depending on specific SHA3 digest length and size of the message. 

For now, this feature will not be enabled automatically for aarch64.  We can auto-enable this when it is fully tested on real hardware. 
But for the above testing purposes, this is auto-enabled when the corresponding hardware feature is detected.

**Comment by RealFYang** [comment]
Bug: https://bugs.openjdk.java.net/browse/JDK-8252204
This added an intrinsic for SHA3 using aarch64 v8.2 SHA3 Crypto Extensions.
Reference implementation for core SHA-3 transform using ARMv8.2 Crypto Extensions:
        https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm64/crypto/sha3-ce-core.S?h=v5.4.52

Trivial adaptation in SHA3. implCompress is needed for the purpose of adding the intrinsic.
For SHA3, we need to pass one extra parameter "digestLength" to the stub for the calculation of block size.
"digestLength" is also used in for the EOR loop before keccak to differentiate different SHA3 variants.

We added jtreg tests for SHA3 and used QEMU system emulator which supports SHA3 instructions to test the functionality. 
Patch passed jtreg tier1-3 tests with QEMU system emulator. 
Also verified with jtreg tier1-3 tests without SHA3 instructions on aarch64-linux-gnu and x86_64-linux-gnu, to make sure that there's no regression. 

We used one existing JMH test for performance test: test/micro/org/openjdk/bench/java/security/MessageDigests.java 
We measured the performance benefit with an aarch64 cycle-accurate simulator.  
Patch delivers 20% - 40% performance improvement depending on specific SHA3 digest length and size of the message. 

For now, this feature will not be enabled automatically for aarch64.  We can auto-enable this when it is fully tested on real hardware. 
But for the above testing purposes, this is auto-enabled when the corresponding hardware feature is detected.

**Comment by RealFYang** [comment]
@ardbiesheuvel : Ard, could you please ack this patch? Thanks.

**Comment by ardbiesheuvel** [comment]
Acked-by: Ard Biesheuvel <ard.biesheuvel@linaro.org>

**Comment by theRealAph** [comment]
> If this feature is not auto-enabled when the SHA3 hardware feature is there, we will have one failure for the following test:
> test/hotspot/jtreg/compiler/intrinsics/sha/cli/TestUseSHA3IntrinsicsOptionOnSupportedCPU.java
> 15 #-----testresult-----
> 16 description=file\:/home/yangfei/github/jdk/test/hotspot/jtreg/compiler/intrinsics/sha/cli/TestUseSHA3IntrinsicsOptionOnSupportedCPU.java
> 17 elapsed=31546 0\:00\:31.546
> 18 end=Mon Sep 21 10\:27\:58 CST 2020
> 19 environment=regtest
> 20 execStatus=Failed. Execution failed\: `main' threw exception\: java.lang.AssertionError\: Option 'UseSHA3Intrinsics' is expected to have 'true' value Option 'UseSHA3Intrinsics' should be enabled by default
> 
> Any suggestions for this?

I don't understand your question. There should be two acceptable results, either "Pass" or "Not supported". What else is possible?

**Comment by valeriepeng** [comment]
I have looked at the java security changes, i.e. src/java.base/share/classes/sun/security/provider/SHA3.java. It looks fine.

**Review by theRealAph — APPROVED** [review]

**Comment by theRealAph** [comment]
I see Linux x64 failed. However, I don't seem to be able to withdraw my patch approval.
However, please consider it withdrawn.

**Comment by RealFYang** [comment]
> I see Linux x64 failed. However, I don't seem to be able to withdraw my patch approval.
> However, please consider it withdrawn.

Thanks for approving this patch.
I checked the error messages and I think the failures were not caused by this patch.
The failures has been fixed by the following two commits:
    commit ec41046c5ce7077eebf4a3c265f79c7fba33d916
    8254348: Build fails when cds is disabled after JDK-8247536
    commit aaa0a2a04792d7c84150e9d972790978ffcc6890
    8254297: Zero and Minimal VMs are broken with undeclared identifier 'DerivedPointerTable' after JDK-8253180

The testing was triggered again automatically after I merge master and I see it passed now.
Do you have any comments for the discussion here?
    https://github.com/openjdk/jdk/pull/207#issuecomment-701243662

Valerie Peng has checked the java security changes, i.e. src/java.base/share/classes/sun/security/provider/SHA3.java. 
Do you think we need another reviewer for this patch?

**Comment by kevinrushforth** [comment]
@theRealAph if you still need to, you can withdraw your approval by reviewing it again and selecting "Changes requested".

**Comment by RealFYang** [comment]
> I have looked at the java security changes, i.e. src/java.base/share/classes/sun/security/provider/SHA3.java. It looks fine.

@valeriepeng : I see you are not listed under "Reviewers" commit message part, could you please press the magic button(s)(approve?) so you get the credit? Thanks.

**Comment by valeriepeng** [comment]
> 
> 
> > I have looked at the java security changes, i.e. src/java.base/share/classes/sun/security/provider/SHA3.java. It looks fine.
> 
> @valeriepeng : I see you are not listed under "Reviewers" commit message part, could you please press the magic button(s)(approve?) so you get the credit? Thanks.

It's fine, the part I reviewed is only a small part of the changes, so I will leave the reviewer approval upto the hotspot team.
Thanks,
Valerie

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Always run graalunit testing with new intrinsics. You need to adjust Graal test:
src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java

**Comment by RealFYang** [comment]
> Always run graalunit testing with new intrinsics. You need to adjust Graal test:
> src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java

Thanks for looking at this. We did run graalunit testing and added the following change in our first commit:

diff --git a/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java      b/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
index f0e17947460..8f3f4ed9323 100644
--- a/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
+++ b/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
@@ -601,6 +601,7 @@ public class CheckGraalIntrinsics extends GraalTest {
          if (!config.useSHA512Intrinsics()) {
              add(ignore, "sun/security/provider/SHA5." + shaCompressName + "([BI)V");
          }
+        add(toBeInvestigated, "sun/security/provider/SHA3." + shaCompressName + "([BI)V");
      }

File: src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
Comment (vnkozlov) [inline src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java:604]: This should be under `if (isJDK16OrHigher())` check. Something like this:
https://github.com/openjdk/jdk/pull/650/files#diff-d1f378fc1b7fe041309e854d40b3a95a91e63fdecf0ecd9826b7c95eaeba314eR527
You can wait when Aleksey push it and update your changes

**Review by vnkozlov — CHANGES_REQUESTED** [review]
Someone in Oracle have to run tier1-tier3 testing with these changes to make sure nothing is broken. I don't want to repeat 8254790.

File: src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
Comment (RealFYang) [reply src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java:604]: OK.  Will update with the following change after Aleksey's PR is integrated:

```
--- a/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
+++ b/src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
@@ -608,6 +608,10 @@ public class CheckGraalIntrinsics extends GraalTest {
         if (!config.useSHA512Intrinsics()) {
             add(ignore, "sun/security/provider/SHA5." + shaCompressName + "([BI)V");
         }
+
+        if (isJDK16OrHigher()) {
+            add(toBeInvestigated, "sun/security/provider/SHA3." + shaCompressName + "([BI)V");
+        }
     }
```

**Comment by RealFYang** [comment]
> Someone in Oracle have to run tier1-tier3 testing with these changes to make sure nothing is broken. I don't want to repeat 8254790.

That's appreciated.
On my side, I run tier1-tier3 both on aarch64 linux and x86_64 linux.
The test result on these two platforms looks good for the latest changes.

File: src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
Comment (vnkozlov) [reply src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java:604]: Yes, please, do that.

**Comment by vnkozlov** [comment]
> > Someone in Oracle have to run tier1-tier3 testing with these changes to make sure nothing is broken. I don't want to repeat 8254790.
> 
> That's appreciated.
> On my side, I run tier1-tier3 both on aarch64 linux and x86_64 linux.
> The test result on these two platforms looks good for the latest changes.

I started testing of 09: version.

File: src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java
Comment (RealFYang) [reply src/jdk.internal.vm.compiler/share/classes/org.graalvm.compiler.hotspot.test/src/org/graalvm/compiler/hotspot/test/CheckGraalIntrinsics.java]: Done.
Commit: https://github.com/openjdk/jdk/pull/207/commits/b43f91970d44e6e0c1b3b4ef452ec388ecbecb83
I think this will not conflict with Aleksey's PR as we modify in different places of CheckGraalIntrinsics.java

**Review by vnkozlov — APPROVED** [review]
tier1,2,3 passed. I verified that new SHA3 tests were run and passed.
But because SHA3 is not enabled for now (even on aarch64), it does not test asm code.
At least testing verified that changes in shared code does not cause any issues.

**Comment by RealFYang** [comment]
> tier1,2,3 passed. I verified that new SHA3 tests were run and passed.
> But because SHA3 is not enabled for now (even on aarch64), it does not test asm code.
> At least testing verified that changes in shared code does not cause any issues.

Great to hear that :-)
Thanks for the effect.
With that testing result and reviewing from three reviewers, I think it's safe to integrate.

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp
Comment (mur47x111) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:3473]: is it intentional to load 16 bytes and post-increment by 8?

Comment (ferakocz) [reply src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:3473]: Actually, with the ld1r instruction the post increment should be the same as the size of the memory accessed. So T2D requires 8 as it reads 8 bytes(and duplicates it into both halves of the SIMD register).

**Pull request closed without merging**
